src/board/system76/common/smfi.c

```diff
@@ -184,33 +184,35 @@ static enum Result cmd_keymap_get(void) {
     int layer = smfi_cmd[2];
     int output = smfi_cmd[3];
     int input = smfi_cmd[4];
-
-    if (layer < KM_LAY && output < KM_OUT && input < KM_IN) {
-        uint16_t key = KEYMAP[layer][output][input];
+    uint16_t key = 0;
+    if (keymap_get(layer, output, input, &key)) {
         smfi_cmd[5] = (uint8_t)key;
         smfi_cmd[6] = (uint8_t)(key >> 8);
         return RES_OK;
+    } else {
+        return RES_ERR;
     }
-
-    // Failed if keyboard mapping not found
-    return RES_ERR;
 }
 
 static enum Result cmd_keymap_set(void) {
     int layer = smfi_cmd[2];
     int output = smfi_cmd[3];
     int input = smfi_cmd[4];
-
-    if (layer < KM_LAY && output < KM_OUT && input < KM_IN) {
-        uint16_t key =
-            ((uint16_t)smfi_cmd[5]) |
-            (((uint16_t)smfi_cmd[6]) << 8);
-        KEYMAP[layer][output][input] = key;
-        return RES_OK;
+    uint16_t key =
+        ((uint16_t)smfi_cmd[5]) |
+        (((uint16_t)smfi_cmd[6]) << 8);
+    //TODO: consider only setting if the key has changed
+    if (keymap_set(layer, output, input, key)) {
+        //TODO: should we write config on every change?
```

Comment:
Maybe something like:
* Keyboard Configurator sends a command to sync to ROM when it closes cleanly
* Sync periodically (every few minutes?) when there are changes
* Sync, if there are changes, on poweroff/suspend

I guess disk caching in an OS like Linux works something like that, though I don't know the exact approach.

Reply:
Indeed. It doesn't seem to cause delays though to keep writing to flash, and it will have the same 100,000 change limitation as Launch has. So if there is an apply button just for preventing flash wear, it should probably happen on launch as well.